File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) 
    {
        if(s.length() < t.length())
        return "";

        unordered_map<char,int> map;
        for(char ch : t)
        map[ch]++;

        int matches = 0, startIndex = -1, minLength = s.length();
        int i = 0;
        for(int j = 0; j < s.length(); j++)
        {
            char ch = s[j];
            if(map.count(ch) > 0)
            {
                map[ch]--;
                if(map[ch] == 0)
                matches++;
            }
            
            while(matches == map.size())
            {
                if(j - i + 1 <= minLength)
                {
                    minLength = j - i + 1;
                    startIndex = i;
                }

                char ch1 = s[i];
                if(map.count(ch1) > 0)
                {
                    map[ch1]++;
                    if(map[ch1] == 1)
                    matches--;
                }
                i++;
            }
        }
        return startIndex == -1 ? "" : s.substr(startIndex, minLength);
    }
};

File: 76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        Map<Character, Integer> map = new HashMap<>();
        for (char c : t.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        int count = map.size(), minLength = s.length();
        int startIndex = -1, i = 0;
        for(int j = 0; j < s.length(); j++){
            char c = s.charAt(j);
            map.put(c, map.getOrDefault(c, 0) - 1);
            if(map.get(c) == 0){
                count--;
            }

            while(count == 0){
                if(j - i + 1 <= minLength){
                    startIndex = i;
                    minLength = Math.min(minLength, j - i + 1);
                }

                c = s.charAt(i);
                map.put(c, map.getOrDefault(c, 0) + 1);
                if(map.get(c) == 1){
                    count++;
                }

                i++;
            }
        }
        return startIndex == -1 ? "" : s.substring(startIndex, startIndex + minLength);
    }
}